core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1756,6 +1762,143 @@ class KafkaController(val config: KafkaConfig,
     }
   }
 
+  def alterIsrs(alterIsrRequest: AlterIsrRequestData, callback: AlterIsrResponseData => Unit): Unit = {
+    val isrsToAlter = mutable.Map[TopicPartition, LeaderAndIsr]()
+
+    alterIsrRequest.topics().forEach(topicReq => topicReq.partitions().forEach(partitionReq => {
+      val tp = new TopicPartition(topicReq.name, partitionReq.partitionIndex)
+      val newIsr = partitionReq.newIsr().asScala.toList.map(_.toInt)
+      isrsToAlter.put(tp, new LeaderAndIsr(partitionReq.leaderId, partitionReq.leaderEpoch, newIsr, partitionReq.currentIsrVersion))
+    }))
+
+    def responseCallback(results: Either[Map[TopicPartition, Errors], Errors]): Unit = {
+      val resp = new AlterIsrResponseData()
+      results match {
+        case Right(error) =>
+          resp.setErrorCode(error.code)
+        case Left(partitions: Map[TopicPartition, Errors]) =>
+          resp.setTopics(new util.ArrayList())
+          partitions.groupBy(_._1.topic).foreachEntry((topic, partitionMap) => {
+            val topicResp = new AlterIsrResponseTopics()
+              .setName(topic)
+              .setPartitions(new util.ArrayList())
+            resp.topics.add(topicResp)
+            partitionMap.foreachEntry((partition, error) => {
+              topicResp.partitions.add(
+                new AlterIsrResponsePartitions()
+                  .setPartitionIndex(partition.partition)
+                  .setErrorCode(error.code))
+            })
+          })
+      }
+      callback.apply(resp)
+    }
+
+    eventManager.put(AlterIsrReceived(alterIsrRequest.brokerId, alterIsrRequest.brokerEpoch, isrsToAlter, responseCallback))
+  }
+
+  private def processAlterIsr(brokerId: Int, brokerEpoch: Long, isrsToAlter: Map[TopicPartition, LeaderAndIsr],
```

Comment:
We should probably have a try/catch in here somewhere for the unhandled errors to make sure that the callback always gets applied.